[Parse] "#if" Conditionals in Collections SR-8743

johnno1962 opened the pull request:
As discussed on the [swift forums](https://forums.swift.org/t/allow-conditional-inclusion-of-elements-in-array-dictionary-literals/16171), a preliminary implementation of conditionals in array and dictionary collections. The basic functionality is straightforward to implement including allowing #sourceLocation directives, more difficult is preserving information in the existing AST model for "-dump-ast" to use for fidelity with the original source for which I’ve put forward a pragmatic implementation using a map. More difficult still is getting lib Syntax to accommodate the change for which I’ll need some help deciphering how it actually works.

See the Jira ticket for more discussion.

Resolves #51252.

johnno1962 committed:
Conditionals in Collections

johnno1962 commented:
@nkcsgexi are you able to help out with this PR at all? I’m trying to adapt collections to allow them to support conditionals using an IfConfigDecl. I’ve worked out a solution of sorts to have the AST dump render conditionals but can’t work out where to begin with the “Syntax” library. Any help appreciated!

nkcsgexi commented:
The most apparent issue we need address is array element in syntax tree has to be an [expression] (https://github.com/apple/swift/blob/master/utils/gyb_syntax_support/ExprNodes.py#L249). We probably need to re-define the ArrayElement node so that it can be either an expression or a IfConfigDecl.

nkcsgexi commented:
Similarly, we need to teach IfConfigDecl to recognize a list of expressions as members, which are not accepted currently ([here](https://github.com/apple/swift/blob/master/utils/gyb_syntax_support/DeclNodes.py#L98)).

johnno1962 commented:
Thanks for the pointers. I tried a few things and couldn’t get swift-syntax-test to work:
```
diff --git a/utils/gyb_syntax_support/DeclNodes.py b/utils/gyb_syntax_support/DeclNodes.py
index 4cf1203bb7..eaa81c2f0a 100644
--- a/utils/gyb_syntax_support/DeclNodes.py
+++ b/utils/gyb_syntax_support/DeclNodes.py
@@ -100,6 +100,8 @@ DECL_NODES = [
                        Child('Statements', kind='CodeBlockItemList'),
                        Child('SwitchCases', kind='SwitchCaseList'),
                        Child('Decls', kind='MemberDeclList'),
+                       Child('ArrayElements', kind='ArrayElementList'),
+#                       Child('DictionaryElements', kind='DictionaryElementList'),
                    ]),
          ]),

diff --git a/utils/gyb_syntax_support/ExprNodes.py b/utils/gyb_syntax_support/ExprNodes.py
index fc94746856..5c18b32420 100644
--- a/utils/gyb_syntax_support/ExprNodes.py
+++ b/utils/gyb_syntax_support/ExprNodes.py
@@ -249,7 +249,11 @@ EXPR_NODES = [
     Node('ArrayElement', kind='Syntax',
          traits=['WithTrailingComma'],
          children=[
-             Child('Expression', kind='Expr'),
+             Child('Expression', kind='Syntax',
+                   node_choices=[
+                       Child('Value', kind='Expr'),
+                       Child('Config', kind='IfConfigDecl'),
+                   ]),
              Child('TrailingComma', kind='CommaToken', is_optional=True),
          ]),
Users-iMac:swift user$
Users-iMac:swift-macosx-x86_64 user$ bin/swift-syntax-test -input-source-filename=d.swift -parse-only
d.swift:132:10: error: unknown expression syntax exists in the source
let a3 = [
         ^
d.swift:134:3: error: unknown declaration syntax exists in the source
  #if false
  ^
```
The main problem is I have no idea what I’m doing when I come to the Syntax code. Does it use the Swift parser and reply on the AST? If so, then the parser works but the AST is not complete without using the IfConfigDecl map I added so something will need to be adapted. Somehow think support for Syntax for the new conditionals in collections should be a separate PR from someone who understands the code!

johnno1962 commented:
AST dump for:
```
let a = [
  1,
  #if true
  2,
  #endif
  3,
]
```
gives:
```
(source_file "/Users/johnholdsworth/d.swift"
  (top_level_code_decl range=[/Users/johnholdsworth/d.swift:2:1 - line:8:1]
    (brace_stmt implicit range=[/Users/johnholdsworth/d.swift:2:1 - line:8:1]
      (pattern_binding_decl range=[/Users/johnholdsworth/d.swift:2:1 - line:8:1]
        (pattern_named type='[Int]' 'a')
        (array_expr type='[Int]' location=/Users/johnholdsworth/d.swift:2:9 range=[/Users/johnholdsworth/d.swift:2:9 - line:8:1]
          (call_expr implicit type='Int' location=/Users/johnholdsworth/d.swift:3:3 range=[/Users/johnholdsworth/d.swift:3:3 - line:3:3] nothrow arg_labels=_builtinIntegerLiteral:
            (constructor_ref_call_expr implicit type='(_MaxBuiltinIntegerType) -> Int' location=/Users/johnholdsworth/d.swift:3:3 range=[/Users/johnholdsworth/d.swift:3:3 - line:3:3] nothrow
              (declref_expr implicit type='(Int.Type) -> (_MaxBuiltinIntegerType) -> Int' location=/Users/johnholdsworth/d.swift:3:3 range=[/Users/johnholdsworth/d.swift:3:3 - line:3:3] decl=Swift.(file).Int.init(_builtinIntegerLiteral:) function_ref=single)
              (type_expr implicit type='Int.Type' location=/Users/johnholdsworth/d.swift:3:3 range=[/Users/johnholdsworth/d.swift:3:3 - line:3:3] typerepr='Int'))
            (tuple_expr implicit type='(_builtinIntegerLiteral: Int2048)' location=/Users/johnholdsworth/d.swift:3:3 range=[/Users/johnholdsworth/d.swift:3:3 - line:3:3] names=_builtinIntegerLiteral
              (integer_literal_expr type='Int2048' location=/Users/johnholdsworth/d.swift:3:3 range=[/Users/johnholdsworth/d.swift:3:3 - line:3:3] value=1)))
          (if_config_decl range=[/Users/johnholdsworth/d.swift:4:3 - line:6:3]
            #if: active
              (boolean_literal_expr type='<null>' value=true)
              (elements
                (integer_literal_expr type='<null>' value=2)))

          (call_expr implicit type='Int' location=/Users/johnholdsworth/d.swift:5:3 range=[/Users/johnholdsworth/d.swift:5:3 - line:5:3] nothrow arg_labels=_builtinIntegerLiteral:
            (constructor_ref_call_expr implicit type='(_MaxBuiltinIntegerType) -> Int' location=/Users/johnholdsworth/d.swift:5:3 range=[/Users/johnholdsworth/d.swift:5:3 - line:5:3] nothrow
              (declref_expr implicit type='(Int.Type) -> (_MaxBuiltinIntegerType) -> Int' location=/Users/johnholdsworth/d.swift:5:3 range=[/Users/johnholdsworth/d.swift:5:3 - line:5:3] decl=Swift.(file).Int.init(_builtinIntegerLiteral:) function_ref=single)
              (type_expr implicit type='Int.Type' location=/Users/johnholdsworth/d.swift:5:3 range=[/Users/johnholdsworth/d.swift:5:3 - line:5:3] typerepr='Int'))
            (tuple_expr implicit type='(_builtinIntegerLiteral: Int2048)' location=/Users/johnholdsworth/d.swift:5:3 range=[/Users/johnholdsworth/d.swift:5:3 - line:5:3] names=_builtinIntegerLiteral
              (integer_literal_expr type='Int2048' location=/Users/johnholdsworth/d.swift:5:3 range=[/Users/johnholdsworth/d.swift:5:3 - line:5:3] value=2)))
          (call_expr implicit type='Int' location=/Users/johnholdsworth/d.swift:7:3 range=[/Users/johnholdsworth/d.swift:7:3 - line:7:3] nothrow arg_labels=_builtinIntegerLiteral:
            (constructor_ref_call_expr implicit type='(_MaxBuiltinIntegerType) -> Int' location=/Users/johnholdsworth/d.swift:7:3 range=[/Users/johnholdsworth/d.swift:7:3 - line:7:3] nothrow
              (declref_expr implicit type='(Int.Type) -> (_MaxBuiltinIntegerType) -> Int' location=/Users/johnholdsworth/d.swift:7:3 range=[/Users/johnholdsworth/d.swift:7:3 - line:7:3] decl=Swift.(file).Int.init(_builtinIntegerLiteral:) function_ref=single)
              (type_expr implicit type='Int.Type' location=/Users/johnholdsworth/d.swift:7:3 range=[/Users/johnholdsworth/d.swift:7:3 - line:7:3] typerepr='Int'))
            (tuple_expr implicit type='(_builtinIntegerLiteral: Int2048)' location=/Users/johnholdsworth/d.swift:7:3 range=[/Users/johnholdsworth/d.swift:7:3 - line:7:3] names=_builtinIntegerLiteral
              (integer_literal_expr type='Int2048' location=/Users/johnholdsworth/d.swift:7:3 range=[/Users/johnholdsworth/d.swift:7:3 - line:7:3] value=3)))
          (semantic_expr
            (call_expr implicit type='[Int]' location=/Users/johnholdsworth/d.swift:2:9 range=[/Users/johnholdsworth/d.swift:2:9 - line:7:3] nothrow arg_labels=arrayLiteral:_:_:
              (constructor_ref_call_expr implicit type='(Int...) -> [Int]' location=/Users/johnholdsworth/d.swift:2:9 range=[/Users/johnholdsworth/d.swift:2:9 - line:2:9] nothrow
                (declref_expr implicit type='(Array<Int>.Type) -> (Int...) -> [Int]' location=/Users/johnholdsworth/d.swift:2:9 range=[/Users/johnholdsworth/d.swift:2:9 - line:2:9] decl=Swift.(file).Array extension.init(arrayLiteral:) [with (substitution_map generic_signature=<Element> (substitution Element -> Int))] function_ref=single)
                (type_expr implicit type='[Int].Type' location=/Users/johnholdsworth/d.swift:2:9 range=[/Users/johnholdsworth/d.swift:2:9 - line:2:9] typerepr='[Int]'))
              (tuple_shuffle_expr implicit type='(arrayLiteral: Int...)' location=/Users/johnholdsworth/d.swift:3:3 range=[/Users/johnholdsworth/d.swift:3:3 - line:7:3] tuple_to_tuple elements=[-2] variadic_sources=[0, 1, 2] default_args_owner=Swift.(file).Array extension.init(arrayLiteral:) [with (substitution_map generic_signature=<Element> (substitution Element -> Int))]
                (tuple_expr implicit type='(arrayLiteral: Int, Int, Int)' location=/Users/johnholdsworth/d.swift:3:3 range=[/Users/johnholdsworth/d.swift:3:3 - line:7:3] names=arrayLiteral,'',''
                  (call_expr implicit type='Int' location=/Users/johnholdsworth/d.swift:3:3 range=[/Users/johnholdsworth/d.swift:3:3 - line:3:3] nothrow arg_labels=_builtinIntegerLiteral:
                    (constructor_ref_call_expr implicit type='(_MaxBuiltinIntegerType) -> Int' location=/Users/johnholdsworth/d.swift:3:3 range=[/Users/johnholdsworth/d.swift:3:3 - line:3:3] nothrow
                      (declref_expr implicit type='(Int.Type) -> (_MaxBuiltinIntegerType) -> Int' location=/Users/johnholdsworth/d.swift:3:3 range=[/Users/johnholdsworth/d.swift:3:3 - line:3:3] decl=Swift.(file).Int.init(_builtinIntegerLiteral:) function_ref=single)
                      (type_expr implicit type='Int.Type' location=/Users/johnholdsworth/d.swift:3:3 range=[/Users/johnholdsworth/d.swift:3:3 - line:3:3] typerepr='Int'))
                    (tuple_expr implicit type='(_builtinIntegerLiteral: Int2048)' location=/Users/johnholdsworth/d.swift:3:3 range=[/Users/johnholdsworth/d.swift:3:3 - line:3:3] names=_builtinIntegerLiteral
                      (integer_literal_expr type='Int2048' location=/Users/johnholdsworth/d.swift:3:3 range=[/Users/johnholdsworth/d.swift:3:3 - line:3:3] value=1)))
                  (call_expr implicit type='Int' location=/Users/johnholdsworth/d.swift:5:3 range=[/Users/johnholdsworth/d.swift:5:3 - line:5:3] nothrow arg_labels=_builtinIntegerLiteral:
                    (constructor_ref_call_expr implicit type='(_MaxBuiltinIntegerType) -> Int' location=/Users/johnholdsworth/d.swift:5:3 range=[/Users/johnholdsworth/d.swift:5:3 - line:5:3] nothrow
                      (declref_expr implicit type='(Int.Type) -> (_MaxBuiltinIntegerType) -> Int' location=/Users/johnholdsworth/d.swift:5:3 range=[/Users/johnholdsworth/d.swift:5:3 - line:5:3] decl=Swift.(file).Int.init(_builtinIntegerLiteral:) function_ref=single)
                      (type_expr implicit type='Int.Type' location=/Users/johnholdsworth/d.swift:5:3 range=[/Users/johnholdsworth/d.swift:5:3 - line:5:3] typerepr='Int'))
                    (tuple_expr implicit type='(_builtinIntegerLiteral: Int2048)' location=/Users/johnholdsworth/d.swift:5:3 range=[/Users/johnholdsworth/d.swift:5:3 - line:5:3] names=_builtinIntegerLiteral
                      (integer_literal_expr type='Int2048' location=/Users/johnholdsworth/d.swift:5:3 range=[/Users/johnholdsworth/d.swift:5:3 - line:5:3] value=2)))
                  (call_expr implicit type='Int' location=/Users/johnholdsworth/d.swift:7:3 range=[/Users/johnholdsworth/d.swift:7:3 - line:7:3] nothrow arg_labels=_builtinIntegerLiteral:
                    (constructor_ref_call_expr implicit type='(_MaxBuiltinIntegerType) -> Int' location=/Users/johnholdsworth/d.swift:7:3 range=[/Users/johnholdsworth/d.swift:7:3 - line:7:3] nothrow
                      (declref_expr implicit type='(Int.Type) -> (_MaxBuiltinIntegerType) -> Int' location=/Users/johnholdsworth/d.swift:7:3 range=[/Users/johnholdsworth/d.swift:7:3 - line:7:3] decl=Swift.(file).Int.init(_builtinIntegerLiteral:) function_ref=single)
                      (type_expr implicit type='Int.Type' location=/Users/johnholdsworth/d.swift:7:3 range=[/Users/johnholdsworth/d.swift:7:3 - line:7:3] typerepr='Int'))
                    (tuple_expr implicit type='(_builtinIntegerLiteral: Int2048)' location=/Users/johnholdsworth/d.swift:7:3 range=[/Users/johnholdsworth/d.swift:7:3 - line:7:3] names=_builtinIntegerLiteral
                      (integer_literal_expr type='Int2048' location=/Users/johnholdsworth/d.swift:7:3 range=[/Users/johnholdsworth/d.swift:7:3 - line:7:3] value=3)))))))))
))
  (var_decl range=[/Users/johnholdsworth/d.swift:2:5 - line:2:5] "a" type='[Int]' interface type='[Int]' access=internal let readImpl=stored immutable))
```

nkcsgexi commented:
@johnno1962 could you also try running `swift-syntax-test` as such: `$swift-syntax-test -input-source-filename /tmp/t.swift -parse-gen -print-node-kind -print-trivial-node-kind`. This will print the whole source file with syntax kind. In the dump, you can find `UnknownSyntax` and see why they're not getting parsed as valid syntax.

johnno1962 committed:
Changes to adapt "Syntax" library

johnno1962 commented:
Thanks @nkcsgexi, that was a huge help! For the following input:
```
let a4 = [
  #if true
    2.0,
    #if false
      3.0,
      #elseif true
      [55, 44],
      #else
      6, 7,
    #endif
  #endif
  32,
]

let d2 = [
  "a": 1,
  #if false
  "b": 2,
  #else
  "bb": 22,
  #endif
  "c": 3,
]
```
“Syntax" works and I now get the following with the last commit applied:
```
let <PatternBindingList><PatternBinding><IdentifierPattern>a4 </IdentifierPattern><InitializerClause>= <ArrayExpr>[<ArrayElementList><IfConfigDecl><IfConfigClauseList><IfConfigClause>
  #if <BooleanLiteralExpr>true</BooleanLiteralExpr><ArrayElementList><ArrayElement><FloatLiteralExpr>
    2.0</FloatLiteralExpr>,</ArrayElement><IfConfigDecl><IfConfigClauseList><IfConfigClause>
    #if <BooleanLiteralExpr>false</BooleanLiteralExpr><ArrayElementList><ArrayElement><FloatLiteralExpr>
      3.0</FloatLiteralExpr>,</ArrayElement></ArrayElementList></IfConfigClause><IfConfigClause>
      #elseif <BooleanLiteralExpr>true</BooleanLiteralExpr><ArrayElementList><ArrayElement><ArrayExpr>
      [<ArrayElementList><ArrayElement><IntegerLiteralExpr>55</IntegerLiteralExpr>, </ArrayElement><ArrayElement><IntegerLiteralExpr>44</IntegerLiteralExpr></ArrayElement></ArrayElementList>]</ArrayExpr>,</ArrayElement></ArrayElementList></IfConfigClause><IfConfigClause>
      #else<ArrayElementList><ArrayElement><IntegerLiteralExpr>
      6</IntegerLiteralExpr>, </ArrayElement><ArrayElement><IntegerLiteralExpr>7</IntegerLiteralExpr>,</ArrayElement></ArrayElementList></IfConfigClause></IfConfigClauseList>
    #endif</IfConfigDecl></ArrayElementList></IfConfigClause></IfConfigClauseList>
  #endif</IfConfigDecl><ArrayElement><IntegerLiteralExpr>
  32</IntegerLiteralExpr>,</ArrayElement></ArrayElementList>
]</ArrayExpr></InitializerClause></PatternBinding></PatternBindingList></VariableDecl></CodeBlockItem><CodeBlockItem><VariableDecl>

let <PatternBindingList><PatternBinding><IdentifierPattern>d2 </IdentifierPattern><InitializerClause>= <DictionaryExpr>[<DictionaryElementList><DictionaryElement><StringLiteralExpr>
  "a"</StringLiteralExpr>: <IntegerLiteralExpr>1</IntegerLiteralExpr>,</DictionaryElement><IfConfigDecl><IfConfigClauseList><IfConfigClause>
  #if <BooleanLiteralExpr>false</BooleanLiteralExpr><DictionaryElementList><DictionaryElement><StringLiteralExpr>
  "b"</StringLiteralExpr>: <IntegerLiteralExpr>2</IntegerLiteralExpr>,</DictionaryElement></DictionaryElementList></IfConfigClause><IfConfigClause>
  #else<DictionaryElementList><DictionaryElement><StringLiteralExpr>
  "bb"</StringLiteralExpr>: <IntegerLiteralExpr>22</IntegerLiteralExpr>,</DictionaryElement></DictionaryElementList></IfConfigClause></IfConfigClauseList>
  #endif</IfConfigDecl><DictionaryElement><StringLiteralExpr>
  "c"</StringLiteralExpr>: <IntegerLiteralExpr>3</IntegerLiteralExpr>,</DictionaryElement></DictionaryElementList>
]</DictionaryExpr></InitializerClause></PatternBinding></PatternBindingList></VariableDecl></CodeBlockItem></CodeBlockItemList>
```

nkcsgexi commented on utils/gyb_syntax_support/ExprNodes.py:
```diff
@@ -23,10 +23,12 @@
     Node('ArrayElementList', kind='SyntaxCollection',
-         element='ArrayElement'),
+         element='Syntax', element_name='Element',
+         element_choices=['ArrayElement', 'IfConfigDecl', 'PoundSourceLocation']),
```
Do we really need `PoundSourceLocation` here?

johnno1962 replied:
That was a suggestion that came up in the Jira https://bugs.swift.org/browse/SR-8743.

nkcsgexi replied:
hmm, ok. If we want to allow #sourcelocation appears in anywhere in the source file, we probably should define a trivia kind for it rather than putting them in the syntax tree structure.

johnno1962 replied:
I’ve amended the commit and raised a new Jira https://bugs.swift.org/browse/SR-8800 for which I’m trying to work out the description. I’m not trying to get #sourceLocation working in every context but just for collections which was quite easy to implement an I can see would be useful for code generators. You’re right though, perhaps Syntax should treat it as Trivia.

nkcsgexi replied:
Thank you! @johnno1962

harlanhaskins commented:
Can you also add a test for this in [test/ModuleInterface/if-configs.swift](https://github.com/apple/swift/blob/master/test/ModuleInterface/if-configs.swift)? I want to make sure the existing algorithm for stripping #if out of inlinable/default argument code still works.

johnno1962 commented:
Sorry @harlanhaskins, when you say “this” I don’t know what you mean. If you can provide a gist with the changes to test/ModuleInterface/if-configs.swift I can test it and add it to the PR.

johnno1962 commented:
Thanks for the test @harlanhaskins and it’s just as well as the module print still contains the conditionals so the test fails:
```
func hazArrayDefaultArg(_ x: [Int] = [
  1, 2,
  #if NOT_PROVIDED
    3,
  #endif
  4,
  #if !NOT_PROVIDED
    5,
  #else
    6,
  #endif
  7
])
```
Does this need to be fixed somewhere, If so any pointers?

harlanhaskins commented:
Yep, that's the issue here. The code in lib/AST/InlinableText.cpp does a walk over the decls and needs to visit each IfConfig. It may be sufficient to override walkToExprPre and explicitly visit the IfConfigs inside of a CollectionExpr.

johnno1962 committed:
Support Module Interface stripping.

harlanhaskins commented:
@johnno1962 Thanks! My concern with this implementation is that it doesn't preserve the existing formatting of the code. That might not matter in practice, but what happens if you just try to call `walkToDeclPre` on each IfConfigDecl in the std::map?

johnno1962 commented:
That would be a very difficult problem to code up. There are Exprs that are only in the Elements of the CollectionExpr and all the Active Elements in the IfConfigDecl would also be in the Elements list and would need to be de-duped then you have nested conditionals… The consumer of this representation is the compiler anyway isn’t it?

harlanhaskins commented:
@johnno1962 Glad it worked! I think the design of #if makes sense -- hoist out the active things and don't walk into it -- but that bites us a lot when we want to actually process them.

harlanhaskins commented on lib/AST/InlinableText.cpp:
```diff
@@ -109,6 +110,15 @@ struct ExtractInactiveRanges : public ASTWalker {
     return false;
   }
 
+  std::pair<bool, Expr *> walkToExprPre(Expr *e) {
```
Can you add a comment explaining the CollectionExpr IfConfig stuff here?

johnno1962 replied:
Sure thing. I’m looking at why recursive conditionals don’t work at the moment. Any ideas?

harlanhaskins replied:
What output are you getting with nested conditionals? It may be a bug in the existing code for stripping #if

mastercmp commented:
tir. 18. sep. 2018 02.12 skrev Xi Ge <notifications@github.com>:

> Similarly, we need to teach IfConfigDecl to recognize a list of
> expressions as members, which are not accepted currently (here
> <https://github.com/apple/swift/blob/master/utils/gyb_syntax_support/DeclNodes.py#L98>
> ).
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/apple/swift/pull/19347#issuecomment-422211019>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/Ahy-b_kLqYvBvxgmC-MKHyltLmb02LCpks5ucDqIgaJpZM4WsXND>
> .
>

rintaro reviewed (changes requested):
Please add tests for invalid cases like:
```
[
#if true
]
#endif
```
```
[
#if true
  42
#endif
]
```
```
[
#if true
  42
#endif
  ,
  12,
]
```
etc, etc...

Also, empty `#if` directives like:
```
[
#if false
#else
  "foo": 42,
#endif
]
```

johnno1962 commented on lib/Parse/ParseExpr.cpp:
```diff
@@ -3218,6 +3219,9 @@ ParserResult<Expr> Parser::parseExprCollection() {
                                 StructureMarkerKind::OpenSquare);
 
+  while (Tok.is(tok::pound_sourceLocation))
+    parseLineDirective();
```
I could but I’d really rather not. I was asked to do this by @dabrahams and it’s so intertwined with this PR I’d need to wait until this PR closes out before I could put it back in.

benlangmuir replied:
I agree with @rintaro.  The changes to `#sourceLocation` are independent and should be in a separate PR.

> could not back out one or other PR independently.

Nothing prevents backing out the `#sourceLocation` change.  The fact that there is a one-way dependency is not a good enough reason to avoid splitting the changes.

Incidentally, the `#sourceLocation` change doesn't appear to be tested.

johnno1962 replied:
I’ll take it out.

benlangmuir replied:
Thanks!

johnno1962 commented on lib/Parse/ParseExpr.cpp:
```diff
@@ -3251,6 +3255,8 @@ ParserResult<Expr> Parser::parseExprCollection() {
     BacktrackingScope Scope(*this);
     auto HasDelayedDecl = State->hasDelayedDecl();
+    while (Tok.is(tok::pound_if))
+      consumeToken();
```
This was a bug. I’ve pushed an amend to change the code to:
```
    while (Tok.is(tok::pound_if) && consumeToken().isValid() &&
           Tok.is(tok::identifier) && consumeToken().isValid());
```

johnno1962 replied:
Thanks for the tip. I’ve pushed the following:
```
    // Skip any initial conditionals.
    while (Tok.is(tok::pound_if)) {
      consumeToken();
      skipUntilTokenOrEndOfLine(tok::unknown);
    }
```

johnno1962 commented on include/swift/Parse/Parser.h:
```diff
@@ -700,11 +700,58 @@ class Parser {
+  /// current AST collection model, a map is generated from the location
+  /// of actual data values to the conditionals that directly preceed them.
+  /// This is transferred to the CollectionExpr when the collection completes
+  /// and is used by the ATSDumper to provide a repsentation of the conditional.
```
If there other important consumers they will need to be adapted like ASTDumper and Syntax. What do you have in mind?

rintaro replied:
`FormatWalker` in `lib/IDE/Formatting.cpp` walks into `IfConfigDecl`. We need to perform code formatting even for inactive `#if` directive blocks.
Eventually, code-formatting can be done by using `libSyntax`, but currently, we are still using `AST`.

johnno1962 replied:
Hmmm. While all elements, active and inactive syntax high OK in the IDE, I’m not sure I can give you a solution on auto-indentation inside the scope of this PR. This is a difficult problem. I’ve updated the comment.

johnno1962 replied:
There is even a FIXME in lib/IDE/Formatting.cpp 
```
        // FIXME: We are going to revisit the behavior and the indentation we
        // want for dictionary/array literals.
```
Better left until #sourceLocation is sorted out and it can be tackled separately.

johnno1962 committed:
Remove handling of #sourceLocation in collections for now.

johnno1962 committed:
Fix for skip of initial conditionals in collection.

rintaro commented on test/Parse/collection_conditionals_error.swift:
```diff
@@ -0,0 +1,46 @@
+  12,
+]
+// expected-error@-5{{expected ',' separator}}
+// expected-error@-4{{expected expression in container literal}}
```
This should be `unexpected ',' separator` error.

johnno1962 replied:
Should be OK now.

rintaro commented on test/Parse/collection_conditionals_error.swift:
```diff
@@ -0,0 +1,46 @@
+]
+// expected-error@-3{{expected ',' separator}}
+// expected-error@-4{{expected expression in container literal}}
+// expected-warning@-9{{expression of type '[String]' is unused}}
```
Could you use trailing `expected-error` style? i.e.
```swift
[ // expected-warning {{expression of type '[String]' is unused}}
   #if FLAG 
   "foo",
   #else
   "foo": 12, // expected-error {{expected ',' separator}}
              // expected-error@-1 {{expected expression in container literal}}
   #endif
 ]
```
This is easier to understand.

johnno1962 replied:
Much better!

johnno1962 committed:
Additional tests, missing comma warning for dictionaries.

johnno1962 commented on test/Parse/collection_conditionals.swift:
```diff
@@ -0,0 +1,44 @@
+// CHECK:        (integer_literal_expr type='<null>' value=22))))
+  #endif
+  "c": 3,
+]
```
This is now tested in collection_conditionals_error.swift

johnno1962 replied:
I’ve adapted then non-error test collection_conditionals.swift to include this.

johnno1962 commented on test/Parse/collection_conditionals_error.swift:
```diff
@@ -0,0 +1,46 @@
+  "foo": 42,
+#endif
+]
+// expected-error@-4{{expected expression in container literal}}
```
Thanks for your comments. I’ve included #else and #ifelse now.

johnno1962 commented on lib/Parse/Parser.cpp:
```diff
@@ -928,19 +929,50 @@ Parser::parseList(tok RightK, SourceLoc LeftLoc, SourceLoc &RightLoc,
+
+  auto IsConditionalsEnd = [&]() -> bool {
+    return Elements && (Tok.is(tok::pound_elseif) ||
+      Tok.is(tok::pound_else) || Tok.is(tok::pound_endif));
```
Thanks!

johnno1962 commented on lib/Parse/Parser.cpp:
```diff
@@ -928,19 +929,50 @@ Parser::parseList(tok RightK, SourceLoc LeftLoc, SourceLoc &RightLoc,
+    if (IfConfigMap) {
+      while (true) {
+        if (Tok.isNot(tok::pound_if))
+          break;
```
Err, yes this was a leftover from processing #sourceLocation

johnno1962 commented on lib/Parse/ParseExpr.cpp:
```diff
@@ -3297,12 +3306,17 @@ ParserResult<Expr> Parser::parseExprArray(SourceLoc LSquareLoc) {
+    }
 
     if (First) {
       if (Tok.isNot(tok::r_square) && Tok.isNot(tok::comma)) {
```
Not sure there is much I can do about the duplicated error message but I’ve kept this block in as we need it to report missing trailing commas but moved it into Parser::parseList so dictionaries are covered as well.

johnno1962 committed:
Report missing commas after first and remove duplicate error.

johnno1962 commented:
Quite a bit of tidying up today. Having your tests set up has made me a bit bolder with the code.

rintaro commented on test/Parse/collection_conditionals_error.swift:
```diff
@@ -0,0 +1,53 @@
+
+[ // expected-warning {{expression of type '[String : Int]' is unused}}
+#if FLAG
+  // FIXME: commenting out until blah blah...
```
Please remove this comment. It's confusing for the test case.

johnno1962 replied:
Done.

rintaro reviewed (commented):
Thanks, It's getting much better! I'll do another round of review in this week.

In the mean time, please add collection literals with `#if` blocks in `test/Syntax/round_trip_parse_gen.swift` and adjust `test/Syntax/Outputs/round_trip_parse_gen.swift.withkinds` accordingly.
`libSyntax` parsing is now a first class citizen in the compiler, and we have to make sure this change doesn't break anything.

johnno1962 committed:
Syntax test + results update.

johnno1962 commented:
I’ve updated the Syntax test and it’s result. Someone should check the canned result though.

rintaro reviewed (commented):
Could you improve diagnostics for missing comma before `#if`?:
```swift
_ = [
  1
#if
  2,
#endif
]
```
Currently this emits:
```
test.swift:2:4: error: expected ']' in container literal expression
  1
   ^
test.swift:1:5: note: to match this opening '['
_ = [
    ^
test.swift:4:4: error: consecutive statements on a line must be separated by ';'
  2,
   ^
   ;
test.swift:4:4: error: expected expression
  2,
   ^
test.swift:4:3: warning: integer literal is unused
  2,
  ^
test.swift:6:1: error: expected expression
]
^
```

johnno1962 committed:
Improve dianostics

johnno1962 commented on lib/Parse/ParseExpr.cpp:
```diff
@@ -3251,6 +3252,13 @@ ParserResult<Expr> Parser::parseExprCollection() {
     auto HasDelayedDecl = State->hasDelayedDecl();
+
+    // Skip any initial conditionals.
+    while (Tok.isAny(tok::pound_if, tok::pound_elseif, tok::pound_else)) {
```
Done.

johnno1962 commented on include/swift/Parse/Parser.h:
```diff
@@ -700,11 +700,59 @@ class Parser {
+  /// Tracking IfConfigDecl for conditionals in collections for AST dump.
+  /// Since there is no place to directly represent conditionals in the
+  /// current AST collection model, a map is generated from the location
+  /// of actual data values to the conditionals that directly preceed them.
```
Oops.

johnno1962 commented on include/swift/Parse/Parser.h:
```diff
@@ -700,11 +700,59 @@ class Parser {
+  /// Nested conditionals are elements of their outer IfConfigDecl structure.
+  class ConfigMap {
+    std::vector <IfConfigDecl *> OuterDecls, *PendingConditionals = &OuterDecls;
+    std::map<const void *,std::vector <IfConfigDecl *>> ConditionalsMap;
```
Done.

johnno1962 commented on lib/Parse/ParseExpr.cpp:
```diff
@@ -3251,6 +3252,13 @@ ParserResult<Expr> Parser::parseExprCollection() {
+    // Skip any initial conditionals.
+    while (Tok.isAny(tok::pound_if, tok::pound_elseif, tok::pound_else)) {
+      consumeToken();
+      skipUntilTokenOrEndOfLine(tok::unknown);
```
If I change it to tok::NUM_TOKENS it breaks another test. Can you think of an alternative.

johnno1962 replied:
Looks like I was wrong there. There test failure was an unrelated change. I’ve pushed an amend.

johnno1962 commented:
I’ve addressed most of your issues and added a test showing the new diagnostic on missing comma..
```
_ = [
  1 // expected-error {{expected ',' separator}}
#if // expected-error {{incomplete condition in conditional compilation directive}}
  2,
#endif // expected-error {{expected expression in container literal}}
 // expected-error@-1 {{unexpected conditional compilation block terminator}}
] // expected-error {{expected expression}}
```

rintaro commented on test/Parse/collection_conditionals_error.swift:
```diff
@@ -0,0 +1,60 @@
+
+_ = [
+  1 // expected-error {{expected ',' separator}}
+#if // expected-error {{incomplete condition in conditional compilation directive}}
```
Sorry I meant `#if true` or something. There's no need to test "missing condition" here.

johnno1962 replied:
Then it works fine. I’ve pushed an amend.

rintaro commented:
@swift-ci Please Test Source Compatibility

johnno1962 committed:
Typos and new key for ConditionalsMap.

johnno1962 commented on include/swift/Parse/Parser.h:
```diff
@@ -700,11 +700,63 @@ class Parser {
+    Expr *registerActiveDataElement(Expr *expr) {
+      if (PendingConditionals == &OuterDecls) {
+        const void *Start = expr->getSourceRange().Start.getOpaquePointerValue();
+        PendingConditionals = &(ConditionalsMap[Start] = OuterDecls);
```
Brilliant! Code updated to use new unsigned ElementIndex key.

johnno1962 commented on include/swift/Parse/Parser.h:
```diff
@@ -700,11 +700,63 @@ class Parser {
+  public:
+    Expr *registerActiveDataElement(Expr *expr) {
+      if (PendingConditionals == &OuterDecls) {
+        const void *Start = expr->getSourceRange().Start.getOpaquePointerValue();
```
This has been removed on moving to new ElementIndex key.

johnno1962 commented on include/swift/Parse/Parser.h:
```diff
@@ -700,11 +700,63 @@ class Parser {
+  /// as well as by libSyntax and to remove conditionals from module interfaces.
+  /// Nested conditionals are elements of their outer IfConfigDecl structure.
+  class ConfigMap {
+    /// OuterDecls - accumlates outer conditionals before seeing active element.
```
So, many, typos.

johnno1962 commented on include/swift/Parse/Parser.h:
```diff
@@ -700,11 +700,63 @@ class Parser {
+    /// OuterDecls - accumlates outer conditionals before seeing active element.
+    /// PendingConditionals - poitner to current vector to add condtionals to.
+    std::vector <IfConfigDecl *> OuterDecls, *PendingConditionals = &OuterDecls;
+    /// ConditionalsMap - Map accumulated of Contiionals preceeding elements.
```
Fixed.

johnno1962 commented:
I’ve made a minor commit to move to using ElementIndex as the key of the ConditionalsMap - removes what was always an unsatisfactory area of the code. The compatibility quite went through OK Debug & Release and the tests seems to be still pending after 7 hours so I’ve pushed this commit now anyway.

rintaro commented:
@swift-ci Please smoke test

rintaro reviewed (commented):
Thanks John! The logic itself looks good to me now!

johnno1962 commented on include/swift/AST/Expr.h:
```diff
@@ -2067,6 +2069,9 @@ class CollectionExpr : public Expr {
   Expr *SemanticExpr = nullptr;
 
+  /// Populated by Parser::ConfigMap to represent conditionals in collections.
+  std::map<unsigned, std::vector <IfConfigDecl *>> ConditionalsMap;
```
It’s a member of CollectionExpr. It would generally be empty unless there were conditionals.

jrose-apple replied:
#19667

rintaro replied:
@jrose-apple Thank you for explanation!

@johnno1962 Sorry, I should have pointed it out earlier.
Could you make this change? If you have any question, feel free to ask me!
To define `ASTContext` managed `ArrayRef`, [`ArchetypeType::setNestedTypes`](https://github.com/apple/swift/blob/e014950170d0500ba340c6d662edf824a4ff57f7/lib/AST/Type.cpp#L2563-L2567) is a good example.

johnno1962 replied:
Thanks for the hint! I’ve pushed a version along the lines you mention if you want to check it.

johnno1962 committed:
Allocate conditionals map in ASTContext.

jrose-apple commented on include/swift/AST/Expr.h:
```diff
@@ -2110,6 +2116,13 @@ class CollectionExpr : public Expr {
+  const MutableArrayRef<ConfigMapPair> &getConditionalsMap() const {
+    return ConditionalsMap;
+  }
+  void setConditionalsMap(const MutableArrayRef<ConfigMapPair> &Map) {
```
Nitpick: since ArrayRef and MutableArrayRef are just a pointer/length pair, we usually pass them around by value. It also seems like these don't need to be Mutable at all—you never mutate them after they're set.

jrose-apple commented on include/swift/AST/Expr.h:
```diff
@@ -29,6 +29,7 @@
 #include "swift/Basic/InlineBitfield.h"
 #include "llvm/Support/TrailingObjects.h"
 #include <utility>
+#include <map>
```
Now unnecessary!

jrose-apple commented on lib/AST/ASTDumper.cpp:
```diff
@@ -2001,21 +2001,43 @@ class PrintExpr : public ExprVisitor<PrintExpr> {
   }
+  void printRelatedIfConfigDecls(unsigned ElementNumber, CollectionExpr *C) {
+    const auto &ConditionalsMap = C->getConditionalsMap();
+    if (ConditionalsMap.size()) {
```
Nitpick: We slightly prefer checking `empty` over `size`.

jrose-apple commented on test/ModuleInterface/if-configs.swift:
```diff
@@ -89,3 +89,39 @@ public func hasClosureDefaultArgWithSinglePoundIf(_ x: () -> Void = {
+// CHECK-NOT:   #if false
+// CHECK-NOT:     50,
+// CHECK-NOT:   #else
+// CHECK:         5,
```
Did a blank line end up in here? That's not the end of the world, but it would be nice to know.

johnno1962 replied:
No reason.. I’ve changed it to a CHECK-NEXT:. 

Thanks for the Review. We must be nearly there. The nits are ready to push once you’ve finished testing.

jrose-apple reviewed (commented):
AST changes look pretty good to me; only minor comments. I'll leave the Parser and Syntax stuff to Rintaro.

jrose-apple commented:
@swift-ci Please Test Source Compatibility

jrose-apple commented:
@swift-ci Please test

johnno1962 commented:
There was a source compatibility failure: https://ci.swift.org/job/swift-PR-source-compat-suite-debug/199//artifact/swift-source-compat-suite/FAIL_fluent_4.2_BuildSwiftPackage.log but it doesn’t seem to be related. I’ve pushed an amend with the nits.

rintaro commented on include/swift/AST/Expr.h:
```diff
@@ -2067,6 +2068,9 @@ class CollectionExpr : public Expr {
   Expr *SemanticExpr = nullptr;
 
+  /// Populated by Parser::ConfigMap to represent conditionals in collections.
+  MutableArrayRef<std::pair<unsigned, IfConfigDecl *>> ConditionalsMap;
```
This can be `ArrayRef` as well.

rintaro commented on include/swift/AST/Expr.h:
```diff
@@ -2110,6 +2115,13 @@ class CollectionExpr : public Expr {
+  const ArrayRef<ConditionalsMapPair> &getConditionalsMap() const {
+    return ConditionalsMap;
+  }
+  void setConditionalsMap(const MutableArrayRef<ConditionalsMapPair> Map) {
```
And this.

rintaro commented on include/swift/Parse/Parser.h:
```diff
@@ -700,11 +700,54 @@ class Parser {
+    }
+
+    void outerConditionalCompletes(IfConfigDecl *ICD) {
+      ConditionalsMap.emplace_back(std::make_pair(PendingIndexNumber, ICD));
```
Could this be just `ConditionalsMap.emplace_back(PendingIndexNumber, ICD)` maybe?

johnno1962 replied:
I have a lot to learn about c++!

johnno1962 committed:
MutableArrayRef -> ArrayRef

rintaro commented:
@swift-ci Please smoke test

johnno1962 commented:
HI @jrose-apple, You’ve added a  swift-evolution-pending-discussion tag to this request. I put it to the forums and Chris said a proposal should not be required. I feel the discussion on the forum pretty much reached conclusion. Are we still on hold?

jrose-apple commented:
Oops, sorry. I was reading @rjmccall's [comment](https://forums.swift.org/t/allow-conditional-inclusion-of-elements-in-array-dictionary-literals/16171/22) to mean that a proposal was still expected. JohnM, is that incorrect?

johnno1962 commented:
Are you asking me? His only mention of proposals was "We don’t want an omnibus proposal saying …"

jrose-apple commented:
No, sorry, "JohnM" not "JohnH".

rjmccall commented:
I think a proposal is required here, yes.  If @lattner disagrees, then the Core Team needs to have a quick discussion first.

johnno1962 commented:
I don’t mind writing something up for the Core Team to make a call on. It’s the three month wait for a review slot and general wasting of everybody’s time for a full review process for such a trivial change I would have thought it was possible to avoid in this case.

rjmccall commented:
We're very conscious of that and are trying to streamline that process as much as possible.  If this does need a proposal, I think we want to make a point of bringing it up for review in the next couple of weeks.

johnno1962 commented:
I’ve filed a proposal https://github.com/apple/swift-evolution/pull/923 but as you might have guessed I hope the “powers that be” sees reason and not put this up for public review. I’d rather keep the powder dry for a more subjective proposal: https://github.com/apple/swift-evolution/pull/917

johnno1962 commented:
Thanks for all your help @rintaro. If we can just get the paperwork done I’m very happy with the result.

CodaFi commented:
@johnno1962 Do you still want us to hold this open?

johnno1962 commented:
Would have been nice, but it’s not particularly important. I’ll close it for you now.
